src/components/embed/contexts/ThemeContext.tsx

```diff
@@ -48,36 +58,50 @@ export const ThemeProvider = ({ children }: { children: ReactNode }) => {
   }, [mode]);
 
   useEffect(() => {
-    const urlParams = new URLSearchParams(window.location.search);
-    const themeParam = urlParams.get("theme") as ThemeMode | null;
+    localStorage.setItem("theme", mode);
 
-    if (themeParam && ["light", "dark", "system"].includes(themeParam)) {
-      setMode(themeParam);
-      return;
+    if (window.parent !== window) {
+      window.parent.postMessage({ type: "THEME_CHANGE", mode, isDarkMode }, "*");
     }
+  }, [mode, isDarkMode]);
 
-    const savedTheme = localStorage.getItem("theme") as ThemeMode | null;
-    if (savedTheme && ["light", "dark", "system"].includes(savedTheme)) {
-      setMode(savedTheme);
-      return;
-    }
+  const lastModeRef = useRef(mode);
+  const setModeTimeoutIDRef = useRef(0);
 
-    setMode("system");
-  }, []);
+  lastModeRef.current = mode;
 
-  useEffect(() => {
-    localStorage.setItem("theme", mode);
+  const handleSetMode = useCallback((mode: ThemeMode) => {
+    const cover = document.getElementById("cover");
 
-    if (window.parent !== window) {
-      window.parent.postMessage({ type: "THEME_CHANGE", mode, isDarkMode }, "*");
+    if (mode === lastModeRef.current || !cover) return;
+
+    const mediaQuery = window.matchMedia("(prefers-color-scheme: dark)");
+    const appliedMode = mediaQuery.matches ? "dark" : "light";
+
+    if (
+      (mode === "system" && appliedMode === lastModeRef.current) ||
+      (lastModeRef.current === "system" && appliedMode === mode)
+    ) {
+      // If mode is (`mode` arg) or was (`lastModeRef.current`) "system" and the
+      // applied theme doesn't change, no need to show the cover, just update the state.
+      setMode(mode);
+
+      return;
     }
-  }, [mode, isDarkMode]);
+
+    window.clearTimeout(setModeTimeoutIDRef.current);
+    cover.removeAttribute("aria-hidden");
```

Comment:
Because some elements have transitions and some other ones don't, changing the theme with the iframe open looks weird, as some colors will change instantly while some other ones will transition. To address that, we should the spinner first and hide it again shortly after (once the transitions are done).